File: week5/InheritanceTypes.cpp
#include<iostream>
using namespace std;

class ClassA
{
    public:
        void aClass()
        {
            cout<<"Iam a ClassA\n";
        }
};

class ClassB
{
    public:
        void bClass()
        {
            cout<<"Iam a ClassB\n";
        }
};

//Multiple Inheritance
class ClassC: public ClassA,public ClassB
{
    public:
        void cClass()
        {
            cout<<"Iam a ClassC\n";
        }
};

//Simple Inheritance
class ClassD: public ClassA
{
    public:
        void dClass()
        {
            cout<<"Iam a ClassD\n";
        }
};


//Hybrid Inheritance
class ClassE: public ClassC,public ClassD
{
    public:
        void eClass()
        {
            cout<<"Iam a ClassE\n";
        }
};

//Multi-Level Inheritance
class ClassF: public ClassD
{
    public:
        void fClass()
        {
            cout<<"Iam a ClassF\n";
        }
};

//Hierarichal Inheritance
class ClassG:public ClassA
{
    public:
        void gClass()
        {
            cout<<"Iam a ClassG\n";
        }
};
class ClassH:public ClassA
{
    public:
        void hClass()
        {
            cout<<"Iam a ClassH\n";
        }
};

int main()
{
    //Simple Inheritance
    cout<<"Simple Inheritance\n";
    ClassD obj1;
    obj1.aClass();
    obj1.dClass();
    //Multiple Inheritance
    cout<<"Multiple Inheritance\n";
    ClassC obj2;
    obj2.aClass();
    obj2.bClass();
    obj2.cClass();
    //Multi-Level Inheritance
    cout<<"Multi-Level Inheritance\n";
    ClassF obj3;
    obj3.aClass();
    obj3.dClass();
    obj3.fClass();
    //Hierarichal Inheritance
    cout<<"Hierarichal Inheritance\n";
    ClassG obj4;
    ClassH obj5;
    obj4.aClass();
    obj4.gClass();
    obj5.aClass();
    obj5.hClass();
    //Hybrid Inheritance
    cout<<"Hybrid Inheritance\n";
    ClassE obj6;
    //obj6.aClass();
    //aClass() method becomes ambiguous here
    obj6.bClass();
    obj6.cClass();
    obj6.dClass();
    obj6.eClass();
    return 0;
}

File: week5/InheritanceTypes.java
class ClassA
{
    public
        void aClass()
        {
            System.out.print("Iam a ClassA"+"\n");
        }
}

class ClassB extends ClassA
{
    public
        void bClass()
        {
            System.out.print("Iam a ClassB"+"\n");
        }
}

class ClassC extends ClassA
{
    public
        void cClass()
        {
            System.out.print("Iam a ClassC"+"\n");
        }
}

class ClassD extends ClassB
{
    public
        void dClass()
        {
            System.out.print("Iam a ClassD"+"\n");
        }
}

/*class ClassE extends ClassA,ClassB
{
    
}*/
/*Java doesnt support Multiple Inheritance and Hybrid Inheritance.*/

public class InheritanceTypes
{
    public static void main(String[] args)
    {
        //Simple Inheritance
        System.out.print("Simple Inheritance"+"\n");
        ClassB class1=new ClassB();
        class1.aClass();
        class1.bClass();
        //Hierarichal Inheritance
        System.out.print("Hierarichal Inheritance"+"\n");
        ClassC class2=new ClassC();
        class1.aClass();
        class1.bClass();
        class2.aClass();
        class2.cClass();
        //Multi-Level Inheritance
        System.out.print("Multi-Level Inheritance"+"\n");
        ClassD class3=new ClassD();
        class3.aClass();
        class3.bClass();
        class3.dClass();
        /*Java does not support multiple inheritances to avoid the diamond problem 
        which causes complexity and ambiguity. Well we can acheive multiple inheritance by the concept of interface.*/
    }
}